Improve the use of both_libraries

bruchar1 opened the pull request:
FIxes #5452 

- Internal dependency objects now carry both_libraries objects, allowing to choose either the static or the shared version of a library, at a later stage. There is now `as_static` and `as_shared` methods on dependencies for that (otherwise the default is automatically used)
- Added a `default_both_libraries` to configure the default behavior of both_dependencies. Previously, it was always the shared version by default.
- Using the same option with 'auto' value, it is now possible to build hierarchies of both_libraries, with static dependencies for static libs, and shared dependencies for shared libs (internal dependencies only).

In my monorepo, I need to compile static and shared versions of many common libs, for different uses. For instance, I usually prefer the shared version, but for plugins, I need to compile them with static dependencies to make them independent from the version of the common libs used in the main app. With the proposed changes, I can remove a lot of duplication in my build files, by using both_libraries instead of having to duplicate shared_library and static_library, along with their respective declare_dependency.

dcbaker commented on mesonbuild/build.py:
```diff
@@ -137,6 +137,12 @@ def get_target_macos_dylib_install_name(ld) -> str:
+
+def extract_targets_as_list(kwargs: T.Dict[str, T.Any], key: str, self_libs: T.List[LibTypes]) -> T.List[LibTypes]:
+    lib_list = listify(kwargs.get(key, [])) + self_libs
+    return [t.get_default_object() if isinstance(t, BothLibraries) else t for t in lib_list]
```
I'm not thrilled with this because the use of `Any` here in the `dict[str, Any]` will prevent the type checker from helping us out, as opposed to just:
```python
def targets_as_list(extra_libs: T.Sequence[LibTypes], self_libs: T.Sequence[LibTypes]) -> T.List[LibTypes]:
    return [t.get_default_object() if isinstance(t, BothLibraries) else t
               for t in itertools.chain(extra_libs, self_libs)]
```

and being used as:
```python
x = targets_as_list(kwargs.get('link_with', []), self.link_targets)
```
This means that if `kwargs` is annotated, that the type checker can see into the function.

dcbaker commented on mesonbuild/build.py:
```diff
@@ -2483,8 +2483,8 @@ def get_default_install_dir(self) -> T.Union[T.Tuple[str, str], T.Tuple[None, No
 class BothLibraries(SecondLevelHolder):
-    def __init__(self, shared: SharedLibrary, static: StaticLibrary) -> None:
-        self._preferred_library = 'shared'
+    def __init__(self, shared: SharedLibrary, static: StaticLibrary, preferred_library: str) -> None:
```
`prefered_library: Literal['shared', 'static', 'auto']`?

dcbaker commented on mesonbuild/build.py:
```diff
@@ -1737,6 +1732,20 @@ def process_vs_module_defs_kw(self, kwargs: T.Dict[str, T.Any]) -> None:
+
+        def _resolve_both_libs(lib: LibTypes) -> LibTypes:
+            if isinstance(lib, BothLibraries):
+                return getattr(lib, bl_type)
```
The use of `getattr` here is going to confuse the type-checker. I think it would be good to just add a method to `BothLibraries` that take the `Literal['static', 'shared', 'auto']` and retuns the value.

dcbaker commented on mesonbuild/build.py:
```diff
@@ -2889,14 +2889,23 @@ class AliasTarget(RunTarget):
         return repr_str.format(self.__class__.__name__, self.get_id())
 
+    @staticmethod
+    def _deps_generator(dependencies: T.Sequence[T.Union[Target, BothLibraries]]) -> T.Generator[Target, None, None]:
```
This can just be `-> T.Iterator[Target]`, to avoid the need for the for the extra parameters

dcbaker commented on mesonbuild/dependencies/base.py:
```diff
@@ -345,6 +345,24 @@ def generate_link_whole_dependency(self) -> Dependency:
+        from ..build import BothLibraries
+
+        new_dep = copy.copy(self)
+        new_dep.libraries = [lib.static if isinstance(lib, BothLibraries) else lib for lib in self.libraries]
```
Maybe to avoid the need for the `isinstance` check here it would be better to add a `as_static` and `as_shared` method directly to the target types so that you can just `new_dep.libraries = [l.as_static() for l in self.libraries]`?

dcbaker reviewed (changes requested):
I have some implementation suggestions, but overall this looks like a nice change to me.

bruchar1 commented:
@dcbaker I think I addressed all your comments. typing in build.py is not easy, as this file is not ready for typing yet, so I didn't tried to run mypy on it...

dcbaker reviewed (approved):
Thanks, looks good to me

dcbaker commented:
>  I think I addressed all your comments. typing in build.py is not easy, as this file is not ready for typing yet, so I didn't tried to run mypy on it...

Yeah. Don't. I've tried to fix it several times and it's a huge amount of work to get it all right. I'm hoping to get back to it after I get the machinefile, options, and coredata done.

xclaesse commented:
I suspect this is breaking gobject-introspection when building GStreamer with default_library=both.

```
FAILED: subprojects/gstreamer/libs/gst/base/GstBase-1.0.gir 
env PKG_CONFIG_PATH=/home/xclaesse/programmation/gstreamer/builddir/meson-uninstalled PKG_CONFIG=/usr/bin/pkg-config 'CC=ccache cc' /usr/bin/x86_64-linux-gnu-g-ir-scanner --quiet --no-libtool --namespace=GstBase --nsversion=1.0 --warn-all --output subprojects/gstreamer/libs/gst/base/GstBase-1.0.gir '--add-init-section=extern void gst_init(gint*,gchar**);g_setenv("GST_REGISTRY_DISABLE", "yes", TRUE);g_setenv("GST_REGISTRY_1.0", "/no/way/this/exists.reg", TRUE);g_setenv("GST_PLUGIN_PATH_1_0", "", TRUE);g_setenv("GST_PLUGIN_SYSTEM_PATH_1_0", "", TRUE);gst_init(NULL,NULL);' --quiet --c-include=gst/base/base.h -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/libs/gst/base -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/libs/gst/base -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/. -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/. -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/libs -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/libs --filelist=/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/libs/gst/base/libgstbase-1.0.so.0.2500.0.p/GstBase_1.0_gir_filelist --include=GLib-2.0 --include=GObject-2.0 --include=GModule-2.0 --include=Gst-1.0 --symbol-prefix=gst --identifier-prefix=Gst --pkg-export=gstreamer-base-1.0 --cflags-begin -DG_DISABLE_DEPRECATED -DG_DISABLE_CAST_CHECKS -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/. -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/. -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/libs -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/libs -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/gst/parse -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/gst/parse -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/gobject-introspection-1.0 --cflags-end -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/. -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/. -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/libs -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/libs -I/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/gst -I/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/gst --add-include-path=/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/gst --add-include-path=/usr/share/gir-1.0 --add-include-path=/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/. --add-include-path=/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/. --add-include-path=/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/libs --add-include-path=/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/libs --add-include-path=/home/xclaesse/programmation/gstreamer/subprojects/gstreamer/gst --add-include-path=/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/gst -L/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/libs/gst/base --library gstbase-1.0 -L/home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer/gst --extra-library=glib-2.0 --extra-library=gobject-2.0 --extra-library=gmodule-2.0 --extra-library=girepository-1.0 --sources-top-dirs /home/xclaesse/programmation/gstreamer/subprojects/gstreamer --sources-top-dirs /home/xclaesse/programmation/gstreamer/builddir/subprojects/gstreamer
(process:927104): GLib-GIO-DEBUG: 17:36:04.441: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
(process:927104): GLib-GIO-DEBUG: 17:36:04.447: _g_io_module_get_default: Found default implementation gvfs (GDaemonVfs) for ‘gio-vfs’
/usr/bin/ld: /home/xclaesse/programmation/gstreamer/builddir/tmp-introspectfhl5josf/GstBase-1.0.o: undefined reference to symbol 'gst_init'
```

xclaesse commented:
The gnome module has a bunch of `isinstance(lib, build.SharedLibrary):` that would break if you get BothLibraries there I think.

xclaesse commented:
@dcbaker @bruchar1 This is causing GStreamer CI regression: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/7753.

bruchar1 commented:
I'm not really setup for working with the gnome module. Would running the test (e.g. framework/7 gnome) with both_libraries trigger the bug? My guess is that calling `isinstance(lib.get('auto'), build.SharedLibrary):` would solve the problem.

Maybe should you open a separate issue to track this?

xclaesse commented:
I managed to write a unit test: https://github.com/mesonbuild/meson/pull/13835.
